### PR TITLE
Make OpaqueUnaryFn pickleable

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138395

Fixes https://github.com/pytorch/pytorch/issues/138070

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: torch/utils/_sympy/functions.py [hunk 1]
```diff
@@ -1200,7 +1200,9 @@ def eval(cls, a):
                 return getattr(sympy, name)(a)
             return None
 
-    OpaqueUnaryFn.__name__ = "OpaqueUnaryFn_" + name
+    nm = "OpaqueUnaryFn_" + name
+    OpaqueUnaryFn.__name__ = nm
+    OpaqueUnaryFn.__qualname__ = nm
```
Comment (Skylion007) [inline torch/utils/_sympy/functions.py:1205]: This works, but should qualname just be the same as name here?

Comment (ezyang) [reply torch/utils/_sympy/functions.py:1205]: I actually tried it with the module qualifier and pickle had the audacity to break GRRR

File: torch/utils/_sympy/functions.py [hunk 2]
```diff
@@ -1200,7 +1200,9 @@ def eval(cls, a):
                 return getattr(sympy, name)(a)
```
Comment (XuehaiPan) [inline torch/utils/_sympy/functions.py:1200]: I think we can lookup only once and cache in the closure for `getattr(math, name)` and `getattr(sympy, name)` to gain some performance.

**Review by XuehaiPan — APPROVED** [review]

**Review by bobrenjc93 — APPROVED** [review]

**Comment by bhack** [comment]
After solving this another one appeared around the corner https://github.com/pytorch/pytorch/issues/139440

**Pull request closed without merging**